exchange/auction_test.go

```diff
@@ -94,92 +112,151 @@ func loadCacheSpec(filename string) (*cacheSpec, error) {
 		roundedPrices:       roundedPrices,
 	}
-	_ = testAuction.doCache(ctx, cache, bids, vast, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
-	found := 0
+	_ = testAuction.doCache(ctx, cache, targData, &specData.BidRequest, 60, &specData.DefaultTTLs, bidCategory)
 
-	for _, cExpected := range specData.ExpectedCacheables {
-		for _, cFound := range cache.items {
-			// make sure Data section matches exactly
-			if !vast {
-				// not testing VAST XML for custom cache key test
-				eq := jsonpatch.Equal(cExpected.Data, cFound.Data)
-				if !eq {
-					continue
-				}
-			}
+	if len(specData.ExpectedCacheables) > len(cache.items) {
+		//t.Errorf("%s:  Less elements were cached than expected \n", fileDisplayName)
+		t.Errorf("%s:  [CACHE_ERROR] Less elements were cached than expected \n", fileDisplayName)
+	} else if len(specData.ExpectedCacheables) < len(cache.items) {
+		t.Errorf("%s:  [CACHE_ERROR] More elements were cached than expected \n", fileDisplayName)
+	} else { // len(specData.ExpectedCacheables) == len(cache.items)
+		var ht map[string]*cacheComparator = make(map[string]*cacheComparator)
+		var formattedData string
+		var compareString string
 
-			// make sure Key value is as expected
-			keymatch := false
-			if len(cExpected.Key) > 0 {
-				// can only verify prefix; remainder is random
-				if strings.HasPrefix(cFound.Key, cExpected.Key) {
-					keymatch = true
-				}
-			} else {
-				if len(cFound.Key) == 0 {
-					// Key is expected to be empty
-					keymatch = true
-				}
+		for _, cExpected := range specData.ExpectedCacheables {
+			formattedData = strings.Replace(string(cExpected.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("%s_%d", formattedData, cExpected.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
 			}
-			if !keymatch {
-				continue
+			ht[compareString].freq += 1
+			if targData.includeCacheVast {
+				ht[compareString].expectedKeys = append(ht[compareString].expectedKeys, cExpected.Key)
 			}
-
-			// make sure TTLSeconds section matches exactly
-			if cExpected.TTLSeconds == cFound.TTLSeconds {
-				found++
+		}
+		for _, cFound := range cache.items {
+			formattedData = strings.Replace(string(cFound.Data), "\\", "", -1)
+			formattedData = strings.Replace(formattedData, " ", "", -1)
+			compareString = fmt.Sprintf("\"%s\"_%d", formattedData, cFound.TTLSeconds)
+			// init cacheComparator element before hashing it. If needed
+			if _, ok := ht[compareString]; !ok {
+				ht[compareString] = &cacheComparator{0, make([]string, 0), make([]string, 0)}
+			}
+			ht[compareString].freq -= 1
+			if targData.includeCacheVast {
+				ht[compareString].actualKeys = append(ht[compareString].actualKeys, cFound.Key)
+			}
+		}
+		for k, cachedElements := range ht {
```

Comment:
Please add brief comments for each of the three for-loops above explaining the purpose of each.  At line 203, we create a map of expected cache values and associated custom cache keys, if any, and the number of times they should occur, at line 216, we examine the actual cache values and decrement the corresponding counters in the map of expected values, and at line 229, we make sure the final counter values are zero, which would mean we got the exact counts we expected.

Reply:
comments added. PR updated